File: backend/src/main/scala/services/battle/microservices/session/api/state/BattleStateEntityResponseApiTypes.scala
package services.battle.microservices.session.api.state

import io.circe.Encoder
import services.battle.microservices.combat.objects.weapon.WeaponKind
import services.battle.microservices.runtime.objects.event.{BattleEventKind, BattleEventParticipant, BattleEventState}
import services.battle.microservices.abilities.objects.pickup.PickupKind
import services.battle.microservices.abilities.objects.pickup.BattlePickupState
import services.battle.microservices.abilities.objects.skill.BattleSlowFieldState

import BattleStateVectorResponse.given

object BattleStateSlowFieldResponse {
  given Encoder[BattleSlowFieldState] =
    Encoder.forProduct7("fieldId", "ownerPlayerId", "ownerHeroId", "position", "radius", "ttlMs", "durationMs")(
      (response: BattleSlowFieldState) =>
        (
          response.fieldId.value,
          response.ownerPlayerId.value,
          response.ownerHeroId.value,
          response.position,
          response.radius.value,
          response.ttlMs.value,
          response.durationMs.value
      )
    )
}

object BattleStateEventParticipantResponse {
  given Encoder[BattleEventParticipant] =
    Encoder.forProduct3("playerId", "heroId", "displayName")((response: BattleEventParticipant) =>
      (response.playerId.value, response.heroId.value, response.displayName.value)
    )
}

object BattleStateEventResponse {
  import BattleStateEventParticipantResponse.given

  given Encoder[BattleEventState] =
    Encoder.forProduct7("eventId", "type", "kind", "elapsedMs", "message", "source", "target")(
      (response: BattleEventState) =>
        (
          response.eventId.value,
          BattleEventKind.wireValue(response.eventKind),
          BattleEventKind.wireValue(response.eventKind),
          response.elapsedMs.value,
          response.message,
          response.source,
          response.target
      )
    )
}

object BattleStatePickupResponse {
  given Encoder[BattlePickupState] =
    Encoder
      .forProduct6("pickupId", "kind", "position", "available", "respawnMs", "weaponKind")(
        (response: BattlePickupState) =>
          (
            response.pickupId.value,
            PickupKind.wireValue(response.pickupKind),
            response.position,
            response.available,
            response.respawnMs.value,
            response.weaponKind.map(WeaponKind.wireValue)
          )
      )
      .mapJson(_.dropNullValues)
}

File: frontend/src/objects/battle/microservices/session/api/state/BattleStateEntityResponseApiTypes.ts
import type { PickupKind } from "../../../abilities/objects/pickup/PickupKind";
import type { WeaponKind } from "../../../combat/objects/weapon/WeaponKind";
import type { BattleEventKind } from "../../../runtime/objects/event/BattleEventKind";
import type { BattleApiVectorDto } from "./BattleStateSharedResponseApiTypes";

export type BattleEventKindDto = BattleEventKind;

export interface BattleStateSlowFieldResponseDto {
  fieldId: string;
  ownerPlayerId: string;
  ownerHeroId: string;
  position: BattleApiVectorDto;
  radius: number;
  ttlMs: number;
  durationMs: number;
}

export interface BattleStatePickupResponseDto {
  pickupId: string;
  kind: PickupKind;
  position: BattleApiVectorDto;
  available: boolean;
  respawnMs: number;
  weaponKind?: WeaponKind;
}

export interface BattleStateEventParticipantResponseDto {
  playerId: string;
  heroId: string;
  displayName: string;
}

export interface BattleStateEventResponseDto {
  eventId: string;
  type: BattleEventKindDto;
  kind: BattleEventKindDto;
  elapsedMs: number;
  message: string;
  source: BattleStateEventParticipantResponseDto;
  target: BattleStateEventParticipantResponseDto;
}

